clang/test/CodeGenSYCL/esimd-vectorize-md.cpp

```diff
@@ -0,0 +1,18 @@
+// RUN: %clang_cc1 -disable-llvm-passes -triple spir64-unknown-unknown-sycldevice \
+// RUN:   -fsycl-is-device -S -emit-llvm %s -o - | FileCheck %s
+
+// This test checks the generation of CMGenxSIMT function attributes
+
+[[intel::sycl_esimd_vectorize(32)]] __attribute__((sycl_device)) void foo1() {}
+// CHECK: @_Z4foo1v() #[[ATTR1:[0-9]+]]
+
+[[intel::sycl_esimd_vectorize(8)]] [[intel::sycl_esimd_vectorize(16)]] __attribute__((sycl_device)) void foo2() {}
+// CHECK: @_Z4foo2v() #[[ATTR2:[0-9]+]]
+
+[[intel::sycl_esimd_vectorize(8)]] __attribute__((sycl_device)) void foo3();
```

Comment:
This isn't a review comment but just for my own curiosity. I would have expected `8` to be applied here since that is the attribute in first declaration of `foo3()` . Off the top of your head, do you know why `16 `is applied @AaronBallman ?

Reply:
That's a good point -- the code in `MergeSYCLIntelESimdVectorizeAttr` should be dropping the second attribute...

Reply:
I think I may have confused you in this -- I thought the code *was* using the initial value because when merging, we would return nullptr for the merged attribute. So I thought the test was wrong, not the Sema code.

I think the new changes are wrong (sorry about that) because the warning is saying that the second attribute is ignored.

Reply:
The warning says that the attribute is already applied with different arguments. It doesn't say which one is ignored.

So, which one should be selected in this case? `16`? Should I revert my recent change then?

Reply:
The location of the diagnostic should make it clear that one is being diagnosed is the second one. See the warning on line 26 of clang/test/SemaSYCL/esimd-vectorize-attr-device.cpp.

I think your most recent change should be reverted and you should debug why CodeGen is seeing the later attribute rather than the earlier one. It may help to look at an -ast-dump of the code just to make sure it looks reasonable.

Reply:
So should I use something like `FD.getPreviousDecl()` in `CGSYCLRuntime.cpp` to get this fixed?

From the AST point of view, everything looks as expected: declaration has the attribute with value `8`, definition has the attribute with value `16`.

Reply:
@AaronBallman , ping

Reply:
Sorry for the delayed response, somehow this review got lost in my queue. :-(

> From the AST point of view, everything looks as expected: declaration has the attribute with value 8, definition has the attribute with value 16.

What I'd expect is that a conflict between values would be an error rather than a warning, to be honest (that's how most attributes work). But if we want it to be a warning instead, I'd expect the AST to look like:

|-FunctionDecl foo3 'void ()'
| `-SYCLIntelESimdVectorizeAttr 8
`-FunctionDecl prev foo3 'void ()'
  `-SYCLIntelESimdVectorizeAttr Inherited 8

e.g., it should inherit the previous attribute and not add a new one.

Reply:
> What I'd expect is that a conflict between values would be an error rather than a warning, to be honest (that's how most attributes work). 
I think conflicting values currently emit a warning for SYCL attributes. If we want to change this, this should be changed for the existing attributes also right?

Anyway, I haven't verified this but from a quick look I think you need to add a `return nullptr` after diagnostic is emitted.

Reply:
I'm OK with both error and warning options, but as @elizabethandrews pointed out, let's keep it as a warning for consistency.

@AaronBallman, how I can reach the AST that you showed for the example below?
```
[[intel::sycl_esimd_vectorize(8)]] __attribute__((sycl_device)) void foo3();
[[intel::sycl_esimd_vectorize(16)]] __attribute__((sycl_device)) void foo3() {}
```

AFAIK, the attribute for the definition of `foo3` is not `Inherited`. It would be `Inherited` if the definition of `foo3` would not have the attribute specified. So, how I could make the definition of `foo3` inherit the attribute from the declaration? Should I examine the previous declaration of `foo3` and skip adding (in `AddSYCLIntelESimdVectorizeAttr`) the attribute if it exists in the declaration already? Would appreciate examples of other attributes which I can copy from.

Reply:
> > What I'd expect is that a conflict between values would be an error rather than a warning, to be honest (that's how most attributes work).
> 
> For SYCL attributes, I think conflicting values currently emit a warning. If we want to change this, this should be changed for the existing attributes also right?

Agreed, and I'm not suggesting a change from warning to error in this patch. I went back and looked again and community is more inconsistent than I had realized, so "that's how most attributes work" was incorrect.

Reply:
> I'm OK with both error and warning options, but as @elizabethandrews pointed out, let's keep it as a warning for consistency.

Agreed.

> @AaronBallman, how I can reach the AST that you showed for the example below?
> 
> ```
> [[intel::sycl_esimd_vectorize(8)]] __attribute__((sycl_device)) void foo3();
> [[intel::sycl_esimd_vectorize(16)]] __attribute__((sycl_device)) void foo3() {}
> ```
> 
> AFAIK, the attribute for the definition of `foo3` is not `Inherited`. It would be `Inherited` if the definition of `foo3` would not have the attribute specified. So, how I could make the definition of `foo3` inherit the attribute from the declaration? Should I examine the previous declaration of `foo3` and skip adding (in `AddSYCLIntelESimdVectorizeAttr`) the attribute if it exists in the declaration already? Would appreciate examples of other attributes which I can copy from.

Oh, interesting, you're right, all of our attributes are doing the right/wrong thing here.

We have a `-Wignored-attributes` warning when there's a duplicate attribute, but we continue to use its value (the diagnostic gives the wrong impression), but the natural consequence of the way we structure our code is to leave the attribute in place on the declaration in which it was found. (`AddSYCLIntelESimdVectorizeAttr()` is called twice, once for each declaration, `MergeSYCLIntelESimdVectorizeAttr()` is called once for the redeclaration but doesn't remove the already-added attribute.)

So I don't think there's anything for you to change here. If we did, it'd be inconsistent with our other attributes, and if we changed those, we'd risk silently changing program behavior for users who are relying on the duplicate attribute behavior. If we're going to break people, I'd rather it be done by turning this into an error rather than a warning. More thought it needed as to what the right thing to do here is, but thankfully, this is pretty low priority as it's an edge case.

Reply:
tl;dr: I'm back to LGTM. :-)

Reply:
@AaronBallman based on what you said, the latest commit in this PR should be reverted right?

Reply:
Yes -- I had missed that there was another commit, but I spotted it now. Good catch, @elizabethandrews!

Reply:
Sure. Thanks, @AaronBallman and @elizabethandrews, I will go ahead and revert the last commit.